lib/active_job/limiter/mixin.rb

```diff
@@ -29,6 +29,68 @@ def limit_queue(expiration:)
             ActiveJob::Limiter.clear_lock_before_perform(job)
           end
         end
+
+        def throttle_job(duration:, extract_resource_id:, metrics_hook: ->(_result, _job) {})
+          active_job_limiter_add_throttle_job_around_enqueue(
+            duration: duration,
+            extract_resource_id: extract_resource_id,
+            metrics_hook: metrics_hook
+          )
+          active_job_limiter_add_throttle_job_around_perform(
+            duration: duration,
+            extract_resource_id: extract_resource_id,
+            metrics_hook: metrics_hook
+          )
+        end
+
+        private
+
+        # Prefixing private methods with active_job_limiter to avoid conflicts with user code
+
+        def active_job_limiter_add_throttle_job_around_enqueue(duration:, extract_resource_id:, metrics_hook:)
+          around_enqueue do |job, block|
+            resource_id = extract_resource_id.call(job)
+
+            if job.instance_variable_get(:@bypass_active_job_limiter_enqueue_locks)
+              block.call
+            elsif ActiveJob::Limiter.acquire_lock_for_job_resource('enqueue', duration, job, resource_id)
+              block.call
+              metrics_hook.call('enqueue.enqueued', job)
+            else
+              job.job_id = nil
+              metrics_hook.call('enqueue.dropped', job)
+            end
+          end
+        end
+
+        def active_job_limiter_add_throttle_job_around_perform(duration:, extract_resource_id:, metrics_hook:)
+          around_perform do |job, block|
+            resource_id = extract_resource_id.call(job)
+
+            if ActiveJob::Limiter.acquire_lock_for_job_resource('perform', duration, job, resource_id)
+              ActiveJob::Limiter.release_lock_for_job_resource('enqueue', job, resource_id)
+              block.call
+              metrics_hook.call('perform.performed', job)
+            elsif ActiveJob::Limiter.acquire_lock_for_job_resource('reschedule', duration, job, resource_id)
+              self.class.send(:active_job_limiter_reschedule_job_for_later, job, duration)
+              metrics_hook.call('perform.rescheduled', job)
+            else
+              job.job_id = nil
+              metrics_hook.call('perform.dropped', job)
+            end
+          end
+        end
+
+        def active_job_limiter_reschedule_job_for_later(existing_job, lock_duration)
+          new_job = existing_job.class.new(*existing_job.arguments)
+
+          # Rescheduled jobs need to pass through around_enqueue without being dropped, so we
```

Comment:
why is this?

Reply:
If we get to this point, that means a different job has already acquired the enqueue lock and the perform lock. When we re-enqueue this job, activejob will once again run it through the around_enqueue hook. So if we don’t set up an escape hatch then this enqueue would always be dropped.

A consequence of this is that even for end users, it is impossible to schedule a job for the future if there is an equivalent job currently enqueued. The original limit_queue also suffers from this flaw. I don’t see it has high priority since it is quite rare for us to schedule future sidekiq jobs at Flexport, and I imagine it would be even more rare that we would do something like that for the kinds of high frequency jobs that this library is meant to help with.